_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+---
+title: "torch for optimization"
+description: > 
+  Torch is not just for deep learning. Its L-BFGS optimizer, complete with Strong-Wolfe line search, is a powerful tool in unconstrained as well as constrained optimization.
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchoptimization
+date: 04-27-2021
+categories:
+  - Torch
+  - R
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+So far, all `torch` use cases we've seen have been in deep learning. However, its automatic differentiation feature is useful in other areas, as well. One prominent example is numerical optimization: We can use `torch` to find the minimum of a function.
+
+In fact, function minimization is *exactly* what happens in training a neural network. But there, the function in question normally is far too complex to even imagine finding its minima analytically. Numerical optimization aims at building up the tools to handle just this complexity. To that end, however, it starts from functions that are far less deeply composed. Instead, they are hand-crafted to pose specific challenges.
+
+This post is a first introduction to numerical optimization with `torch`. Central takeaways are the existence and usefulness of its L-BFGS optimizer, as well as the impact of running L-BFGS with line search. As a fun add-on, we show an example of constrained optimization, where a constraint is enforced via a quadratic penalty function.
+
+To warm up, we take a detour, minimizing a function "ourselves" using nothing but tensors. This will turn out to be less of a detour than expected though: Later, the overall process will still be the same. All changes will be related to integration of `optimizer`s and their capabilities.
```

Comment:
```suggestion
To warm up, we take a detour, minimizing a function "ourselves" using nothing but tensors. This will turn out to be relevant later, though, as the overall process will still be the same. All changes will be related to integration of `optimizer`s and their capabilities.
```

Reply:
What is `optimizer`? You use backticks around R package names in your posts. Backticks also go around code (e.g. functions and/or their arguments—though you should have trailing parentheses if it's a function).

Reply:
Yes, I see what you mean... I think I'm doing it because I'm not too sure about where to place "optimizer", ontology-wise ... If I used it without any formatting, it would be as though it existed in natural language, which I think it does not, or not really. So it's a technical term that signifies a very specific thing, in deep learning, while in other areas (like databases), it refers to something totally different ... 

So by formatting it like code I'm kinda referring to a virtual superclass - that does in fact exist, but is not named that way - subsuming objects created by `optim_adam()` etc. ..

I don't really like it, but I don't know of a better way (yet)...